File: src/network/packet/ResponseClientPacket.php
<?php

namespace pocketcloud\cloud\network\packet;

use pocketcloud\cloud\network\packet\util\PacketData;

/**
 * A different version from the regular ResponsePacket
 * This logic is reversed, means the sub-servers sends this ResponseClientPacket in response to the RequestClientPacket
 * @see RequestClientPacket
 * @see ResponseClientPacket
 */
abstract class ResponseClientPacket extends CloudPacket implements CloudboundPacket {

    private string $requestId = "";

    final public function encode(PacketData $packetData): void {
        parent::encode($packetData);
        $packetData->write($this->requestId);
    }

    final public function decode(PacketData $packetData): void {
        parent::decode($packetData);
        $this->requestId = $packetData->readString();
    }

    final public function encodePayload(PacketData $packetData): void {}

    public function getRequestId(): string {
        return $this->requestId;
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/ResponseClientPacket.java
package de.pocketcloud.cloud.network.packet;

import de.pocketcloud.cloud.network.packet.util.PacketData;
import lombok.Getter;

/**
 * The reversed response: sub-servers send this back to the cloud in response to a RequestClientPacket.
 * Logic is reversed compared to the regular ResponsePacket — here the cloud is the receiver.
 * @see RequestClientPacket
 */
@Getter
public abstract class ResponseClientPacket extends CloudPacket implements CloudboundPacket {

    private String requestId = "";

    @Override
    public final void encode(PacketData packetData) {
        super.encode(packetData);
        packetData.write(requestId);
    }

    @Override
    public final void decode(PacketData packetData) {
        super.decode(packetData);
        requestId = packetData.readString();
    }

    @Override
    public final void encodePayload(PacketData packetData) {}
}